zerver/lib/url_preview/preview.py

```diff
@@ -14,10 +14,8 @@
 from zerver.lib.url_preview.oembed import get_oembed_data
 from zerver.lib.url_preview.parsers import GenericParser, OpenGraphParser
 
-# FIXME: Should we use a database cache or a memcached in production? What if
-# OpenGraph data is changed for a site?
 # Use an in-memory cache for development, to make it easy to develop this code
-CACHE_NAME = "database" if not settings.DEVELOPMENT else "in-memory"
+CACHE_NAME = "default" if not settings.DEVELOPMENT else "in-memory"
```

Comment:
So ... I think the one possible bug that comes with this change is that memcached is not persistent storage, which means sometimes the code using this cache will fail if it races with a memcached flush:

```
@assign_queue("embed_links")                                                                             
class FetchLinksEmbedData(QueueProcessingWorker):                                                        
    # This is a slow queue with network requests, so a disk write is negligible.                         
    # Update stats file after every consume call.                                                        
    CONSUME_ITERATIONS_BEFORE_UPDATE_STATS_NUM = 1                                                       
                                                                                                         
    def consume(self, event: Mapping[str, Any]) -> None:                                                 
        for url in event["urls"]:                                                                        
            start_time = time.time()                                                                     
            url_preview.get_link_embed_data(url)                                                         
            logging.info(                                                                                
                "Time spent on get_link_embed_data for %s: %s", url, time.time() - start_time            
            )                                                                                            
                                                                                                         
        message = Message.objects.get(id=event["message_id"])                                            
        # If the message changed, we will run this task after updating the message                       
        # in zerver.actions.message_edit.check_update_message                                            
        if message.content != event["message_content"]:                                                  
            return                                                                                       
        if message.content is not None:                                                                  
            query = UserMessage.objects.filter(                                                          
                message=message.id,                                                                      
            )                                                                                            
            message_user_ids = set(query.values_list("user_profile_id", flat=True))                      
                                                                                                         
            # Fetch the realm whose settings we're using for rendering                                   
            realm = Realm.objects.get(id=event["message_realm_id"])                                      
                                                                                                         
            # If rendering fails, the called code will raise a JsonableError.                            
            rendering_result = render_incoming_message(                                                  
                message, message.content, message_user_ids, realm                                        
            )                                                                                            
            do_update_embedded_data(message.sender, message, message.content, rendering_result)          
```

Reading that code ... it's the same process that is doing the rendering as which did the URL lookups, so in theory we could just re-architect further to actually just have that first loop collect a dictionary of results, that is then just passed to `render_incoming_message` as an optional parameter.

(And even if they weren't going to be, we could just include the results dictionary in what we send to that other process).

Reply:
Though now that I think about it, there may be a problem with my proposal, which is that I bet that reason to have a cache is to help avoid performance issues when folks edit a message containing one of those links, so maybe we do still want memcached? But I think we could still have the caching be entirely inside `url_preview.get_link_embed_data`, and be purely a cache, not an IPC mechanism, and then have the Markdown processor itself still just receive a dictionary mapping URLs to embedding data as a parameter to `render_incoming_message`.